### DISCUSSION TITLE
[Feature] Allow specifying 'preid' and 'skipChangelog' on 'nx release' command

### DISCUSSION THREAD

**Discussion by roman-supy-io** [nrwl/nx#28613] [Feature Requests]
[Motivation]
Sometimes you don't have to generate changelogs and sometimes you do.
In our case it depends on the branch.
`nx release` already contains `skipPublish` flag.

[Current solution]
Use custom script and `releaseVersion` `releaseChangelog` and `releasePublish` separately which makes it dirty to copy-paste throughout repositories. The above-mentioned commands cannot be run subsequently also.

**Comment by tristanzander** [comment]
There are separate commands for `nx release version`, `nx release changelog`, and `nx release publish`. Do any of those resolve your use-case and allow you to orchestrate via means other than your custom script?

**Reply by roman-supy-io** [reply]
partially, because `changelog` command requires version, which is not known if all of the three commands are run separately.
any ideas?

**Reply by tristanzander** [reply]
Ah yeah that makes sense then. I don't see why the `nx release` command shouldn't have a `--skip-changelog` argument in this case.

I know there are other commands to print the project graph or just names of affected modules for CI automation, although I don't know it well enough to have a solution for your use-case. Could you pull the script from some other source so that you don't have to copy-paste it in every repo? Honestly, a custom script doesn't sound like a bad idea for your use-case

**Reply by roman-supy-io** [reply]
true, I that's not a complex thing to do, but just the effort of publishing/maintaining everything instead f having this OOB is the only reason why I don't really like it
hope someone from the NX team can look into this and add the `--skip-changelog` and `--preid` flags

**Comment by JamesHenry** [comment]
Why is it dirty to copy paste a script but not CLI commands? 😅 The programmatic API is a first class part of `nx release` and I use it personally as I love to have control, type-checking, intellisense etc. It also allows exactly what you are asking for in terms of changing behaviour conditionally based on branch or anything else from the environment.

What does "The above-mentioned commands cannot be run subsequently also." mean please?

**Reply by roman-supy-io** [reply]
@JamesHenry that means: changelog step requires version step output, which obviously cannot be passed when using CLI approach

**Reply by roman-supy-io** [reply]
I mean, if the flexibility is a thing here, why not make it REALLY flexible and not just call it so? :)

**Reply by JamesHenry** [reply]
Oh I see, you referenced the node API function names in your post, rather than CLI commands, so that is why I was confused.

> I mean, if the flexibility is a thing here, why not make it REALLY flexible and not just call it so? :)

Again, I am referring to the API but I think you are talking about the CLI here? What about the API specifically is not flexible?

**Reply by JamesHenry** [reply]
```ts
// Only generate changelogs when the branch is a particular name
if (process.env.SOME_BRANCH_NAME_ENV_VAR === 'special-branch') {
  await releaseChangelog({ ... })
}
```

**Reply by JamesHenry** [reply]
CLI surface areas can _explode_ if every custom use-case is attempted to be supported directly, and it becomes very hard to maintain. Having the programmatic API in as a first class use-case from day one was an intentional design decision to allow for custom use-cases to be easily and well supported without the more simplistic cases being lost in a sea of CLI flags.

I'm not looking to dismiss the request out of hand, but this is the background and context.

I'd love to understand a bit better why the programmatic API is not a good fit for this use case you have please
